src/coreclr/vm/jitinterface.cpp

```diff
@@ -10623,10 +10631,41 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTINTERFACE ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS ||
             dynamicFtnNum == DYNAMIC_CORINFO_HELP_CHKCASTCLASS_SPECIAL ||
-            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX)
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_UNBOX ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_ARRADDR_ST ||
+            dynamicFtnNum == DYNAMIC_CORINFO_HELP_LDELEMA_REF)
         {
             Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);
             _ASSERTE(pPrecode->GetType() == PRECODE_FIXUP);
+            
+            MethodDesc* helperMD = pPrecode->GetMethodDesc();
+            _ASSERT(helperMD != nullptr);
+
+            // Check if the target MethodDesc is already jitted to its final Tier
+            // so we no longer need to use indirections and can emit a direct call instead.
+            NativeCodeVersion::OptimizationTier tier;
+
+            // Avoid taking the lock for foreground jit compilations
+            if (TieredCompilationManager::IsBackgroundWorkerActive())
```

Comment:
The background worker would probably be created fairly early, and lock contention with that is probably not as interesting. The background worker would be inactive most of the time after startup. There can be multiple threads during startup doing foreground jitting. Maybe a better check would be the following, which would avoid the locked path during startup-like activities:

```suggestion
            if (!GetAppDomain()->GetTieredCompilationManager()->IsTieringDelayActive())
```